mujoco_warp/_src/sensor.py

```diff
@@ -1254,3 +1256,227 @@ def sensor_acc(m: Model, d: Data):
+  jnt_type: wp.array(dtype=int),
+  jnt_qposadr: wp.array(dtype=int),
+  jnt_stiffness: wp.array(dtype=float),
+  # Data in:
+  qpos_in: wp.array2d(dtype=float),
+  # Data out:
+  energy_out: wp.array(dtype=wp.vec2),
+):
+  worldid, jntid = wp.tid()
+  stiffness = jnt_stiffness[jntid]
+
+  if stiffness == 0.0:
+    return
+
+  padr = jnt_qposadr[jntid]
+  jnttype = jnt_type[jntid]
+
+  if jnttype == int(JointType.FREE.value):
+    dif0 = wp.vec3(
+      qpos_in[worldid, padr + 0] - qpos_spring[padr + 0],
+      qpos_in[worldid, padr + 1] - qpos_spring[padr + 1],
+      qpos_in[worldid, padr + 2] - qpos_spring[padr + 2],
+    )
+
+    # convert quaternion difference into angular "velocity"
+    quat1 = wp.quat(
+      qpos_in[worldid, padr + 3],
+      qpos_in[worldid, padr + 4],
+      qpos_in[worldid, padr + 5],
+      qpos_in[worldid, padr + 6],
+    )
+    quat1 = wp.normalize(quat1)
+
+    quat_spring = wp.quat(
+      qpos_spring[padr + 3],
+      qpos_spring[padr + 4],
+      qpos_spring[padr + 5],
+      qpos_spring[padr + 6],
+    )
+
+    dif1 = math.quat_sub(quat1, quat_spring)
+
+    energy = wp.vec2(
+      0.5 * stiffness * (wp.dot(dif0, dif0) + wp.dot(dif1, dif1)),
+      0.0,
+    )
+
+    wp.atomic_add(energy_out, worldid, energy)
+
+  elif jnttype == int(JointType.BALL.value):
+    quat = wp.quat(
+      qpos_in[worldid, padr + 0],
+      qpos_in[worldid, padr + 1],
+      qpos_in[worldid, padr + 2],
+      qpos_in[worldid, padr + 3],
+    )
+    quat = wp.normalize(quat)
+
+    quat_spring = wp.quat(
+      qpos_spring[padr + 0],
+      qpos_spring[padr + 1],
+      qpos_spring[padr + 2],
+      qpos_spring[padr + 3],
+    )
+
+    dif = math.quat_sub(quat, quat_spring)
+    energy = wp.vec2(
+      0.5 * stiffness * wp.dot(dif, dif),
+      0.0,
+    )
+    wp.atomic_add(energy_out, worldid, energy)
+  elif jnttype == int(JointType.SLIDE.value) or jnttype == int(JointType.HINGE.value):
+    dif_ = qpos_in[worldid, padr] - qpos_spring[padr]
+    energy = wp.vec2(
+      0.5 * stiffness * dif_ * dif_,
+      0.0,
+    )
+    wp.atomic_add(energy_out, worldid, energy)
+
+
+@wp.kernel
+def _energy_pos_passive_tendon(
+  # Model:
+  tendon_stiffness: wp.array(dtype=float),
+  tendon_lengthspring: wp.array(dtype=wp.vec2),
+  # Data in:
+  ten_length_in: wp.array2d(dtype=float),
+  # Data out:
+  energy_out: wp.array(dtype=wp.vec2),
+):
+  worldid, tenid = wp.tid()
+
+  stiffness = tendon_stiffness[tenid]
+
+  if stiffness == 0.0:
+    return
+
+  length = ten_length_in[worldid, tenid]
+
+  # compute spring displacement
+  lengthspring = tendon_lengthspring[tenid]
+  lower = lengthspring[0]
+  upper = lengthspring[1]
+
+  if length > upper:
+    displacement = upper - length
+  elif length < lower:
+    displacement = lower - length
+  else:
+    displacement = 0.0
+
+  energy = wp.vec2(0.5 * stiffness * displacement * displacement, 0.0)
+  wp.atomic_add(energy_out, worldid, energy)
+
+
+def energy_pos(m: Model, d: Data):
```

Comment:
`energy_pos` is run inside `forward` as long as `mjENABLED(mjENBL_ENERGY)` - should we do the same in mjwarp?

Reply:
added todos to `forward`